File: smash-ranks-android/app/src/main/java/com/garpr/android/activities/PlayersActivity.java
package com.garpr.android.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.SearchView;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;

import com.garpr.android.App;
import com.garpr.android.R;
import com.garpr.android.misc.RegionManager;
import com.garpr.android.misc.SearchQueryHandle;
import com.garpr.android.models.PlayersBundle;
import com.garpr.android.views.PlayersLayout;

import javax.inject.Inject;

import butterknife.BindView;

public class PlayersActivity extends BaseActivity implements MenuItemCompat.OnActionExpandListener,
        PlayersLayout.Listener, SearchQueryHandle, SearchView.OnQueryTextListener {

    private static final String TAG = "PlayersActivity";

    private SearchView mSearchView;

    @Inject
    RegionManager mRegionManager;

    @BindView(R.id.playersLayout)
    PlayersLayout mPlayersLayout;


    public static Intent getLaunchIntent(final Context context) {
        return getLaunchIntent(context, null);
    }

    public static Intent getLaunchIntent(final Context context, @Nullable final String region) {
        final Intent intent = new Intent(context, PlayersActivity.class);

        if (!TextUtils.isEmpty(region)) {
            intent.putExtra(EXTRA_REGION, region);
        }

        return intent;
    }

    @Override
    protected String getActivityName() {
        return TAG;
    }

    @Nullable
    @Override
    public CharSequence getSearchQuery() {
        return mSearchView == null ? null : mSearchView.getQuery();
    }

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.get().getAppComponent().inject(this);
        setContentView(R.layout.activity_players);
        setSubtitle(mRegionManager.getRegion(this));
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.activity_players, menu);

        final PlayersBundle playersBundle = mPlayersLayout.getPlayersBundle();
        if (playersBundle != null && playersBundle.hasPlayers()) {
            final MenuItem searchMenuItem = menu.findItem(R.id.miSearch);
            searchMenuItem.setVisible(true);

            MenuItemCompat.setOnActionExpandListener(searchMenuItem, this);
            mSearchView = (SearchView) MenuItemCompat.getActionView(searchMenuItem);
            mSearchView.setQueryHint(getText(R.string.search_players_));
            mSearchView.setOnQueryTextListener(this);
        }

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onMenuItemActionCollapse(final MenuItem item) {
        search(null);
        return true;
    }

    @Override
    public boolean onMenuItemActionExpand(final MenuItem item) {
        return true;
    }

    @Override
    public void onPlayersBundleFetched(final PlayersLayout layout) {
        supportInvalidateOptionsMenu();
    }

    @Override
    public boolean onQueryTextChange(final String newText) {
        search(newText);
        return false;
    }

    @Override
    public boolean onQueryTextSubmit(final String query) {
        search(query);
        return false;
    }

    private void search(@Nullable final String query) {
        mPlayersLayout.search(query);
    }

    @Override
    protected boolean showUpNavigation() {
        return true;
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/activities/PlayersActivity.kt
package com.garpr.android.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.widget.SearchView
import com.garpr.android.R
import com.garpr.android.extensions.appComponent
import com.garpr.android.extensions.putOptionalExtra
import com.garpr.android.extensions.subtitle
import com.garpr.android.managers.RegionManager
import com.garpr.android.misc.SearchQueryHandle
import com.garpr.android.misc.Searchable
import com.garpr.android.models.Region
import com.garpr.android.views.PlayersLayout
import com.garpr.android.views.toolbars.SearchToolbar
import kotlinx.android.synthetic.main.layout_players.*
import javax.inject.Inject

class PlayersActivity : BaseActivity(), MenuItem.OnActionExpandListener, PlayersLayout.Listener,
        Searchable, SearchQueryHandle, SearchToolbar.Listener, SearchView.OnQueryTextListener {

    private var searchView: SearchView? = null

    @Inject
    protected lateinit var regionManager: RegionManager


    companion object {
        private const val TAG = "PlayersActivity"

        fun getLaunchIntent(context: Context, region: Region? = null): Intent {
            return Intent(context, PlayersActivity::class.java)
                    .putOptionalExtra(EXTRA_REGION, region)
        }
    }

    override val activityName = TAG

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        appComponent.inject(this)
        setContentView(R.layout.activity_players)
        subtitle = regionManager.getRegion(this).displayName
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_players, menu)

        val searchMenuItem = menu.findItem(R.id.miSearch) ?: throw RuntimeException(
                "searchMenuItem is null")
        searchMenuItem.isVisible = showSearchMenuItem
        searchMenuItem.setOnActionExpandListener(this)

        val searchView = searchMenuItem.actionView as SearchView
        searchView.queryHint = getText(R.string.search_)
        searchView.setOnQueryTextListener(this)
        this.searchView = searchView

        return super.onCreateOptionsMenu(menu)
    }

    override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
        search(null)
        return true
    }

    override fun onMenuItemActionExpand(item: MenuItem) = true

    override fun onPlayersBundleFetched(layout: PlayersLayout) = invalidateOptionsMenu()

    override fun onQueryTextChange(newText: String): Boolean {
        search(newText)
        return false
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        search(query)
        return false
    }

    override fun search(query: String?) = playersLayout.search(query)

    override val searchQuery: CharSequence?
        get() = searchView?.query

    override val showSearchMenuItem: Boolean
        get() = playersLayout.playersBundle?.players?.isNotEmpty() == true

}
